romanisim/tests/test_linear.py

```diff
@@ -0,0 +1,185 @@
+    lin_coeffs = np.tile(coeffs[:, np.newaxis, np.newaxis], (1, 100, 100))
+
+    lin_coeffs[:, 0:50, :] *= 2.0
+    rev_lin_coeffs = lin_coeffs[::-1, ...]
+    gain = 4.0 * u.electron / u.DN
+
+    linearity = nonlinearity.NL(lin_coeffs, gain)
+    rev_linearity = nonlinearity.NL(rev_lin_coeffs, gain)
+
+    res = linearity.apply(counts)
+    res_rev = rev_linearity.apply(counts, reversed=True)
+
+    assert np.all(res_rev[:] == res[:])
+
+@pytest.mark.skipif(
+    os.environ.get("CI") == "true",
+    reason=(
+        "Roman CRDS servers are not currently available outside the internal network"
+    ),
+)
+def test_inverse_then_linearity():
+    # Test that applying inverse linearity and then linearity returns the results to
+    # the original value
+
+    # Set up parameters for simulation run
+    galsim.roman.n_pix = 4088
+    metadata = copy.deepcopy(parameters.default_parameters_dictionary)
+    metadata['instrument']['detector'] = 'WFI07'
+    metadata['instrument']['optical_element'] = 'F158'
+    metadata['exposure']['ma_table_number'] = 1
+
+    twcs = wcs.get_wcs(metadata, usecrds=True)
+    rd_sca = twcs.toWorld(galsim.PositionD(
+        galsim.roman.n_pix / 2, galsim.roman.n_pix / 2))
+
+    cat = catalog.make_dummy_table_catalog(
+        rd_sca, bandpasses=[metadata['instrument']['optical_element']], nobj=1000)
+
+    rng = galsim.UniformDeviate(None)
+
+    meta = maker_utils.mk_common_meta()
+    meta["photometry"] = maker_utils.mk_photometry()
+
+    for key in parameters.default_parameters_dictionary.keys():
+        meta[key].update(parameters.default_parameters_dictionary[key])
+
+    util.add_more_metadata(meta)
+
+    for key in metadata.keys():
+        meta[key].update(metadata[key])
+
+    image_node = maker_utils.mk_level2_image()
+    image_node['meta'] = meta
+    image_mod = roman_datamodels.datamodels.ImageModel(image_node)
+
+    refnames_lst = ['inverselinearity', 'linearity']
+
+    reffiles = crds.getreferences(
+        image_mod.get_crds_parameters(), reftypes=refnames_lst,
+        observatory='roman')
+
+    inverse_linearity_model = roman_datamodels.datamodels.InverselinearityRefModel(
+        reffiles['inverselinearity'])
+    linearity_model = roman_datamodels.datamodels.LinearityRefModel(
+        reffiles['linearity'])
+
+    gain = 1.0
+
+    inverse_linearity = nonlinearity.NL(
+        inverse_linearity_model.coeffs[:,0:4088,0:4088], gain=gain)
+    linearity = nonlinearity.NL(
+        linearity_model.coeffs[:,0:4088,0:4088], gain=gain)
+
+
+
+    level_0, simcatobj = image.simulate(
+        metadata, cat, usecrds=True,
+        webbpsf=True, level=0,
+        rng=rng)
+
+    level_0_inv = inverse_linearity.apply(level_0['data'])
+
+    level_0_lin = linearity.apply(level_0_inv)
+
+    # Divide initial counts before and after inverse & NL application
+    div = level_0['data'][np.isfinite(level_0['data'])] / level_0_lin[np.isfinite(level_0_lin)]
```

Comment:
This is a good test, but there's no reason to use a real image for it?  i.e., something like:
```
counts = np.poisson(...)
nl_counts = inverse_linearity.apply(counts)
assert np.all(counts == lineary.apply(nl_counts))
```
Then you can get rid of all of the bits of the test that go into making that level 0 file (catalogs, metadata, ...).

Reply:
Trimmed down.